Improve hover results

HedgehogCode opened the pull request:
This PR solves #444.

- Use jedi Script.help instead of infer - Ensures that docstrings of the definition are used if possible instead of docstrings of the type
- Show all possible signatures or types of the element hovered instead of just the first one that matches the name

I tried to describe the changes pretty well in the comments in the code. Note that there are two `TODO(Review)` comments that I would like feedback on.

Here are some examples of how the hover results change with this code change. The new results are highlighted with a green border.

![hover_preview_1](https://github.com/python-lsp/python-lsp-server/assets/6631116/49027efb-b0be-46af-ab0c-58da39b33605)
![hover_preview_2](https://github.com/python-lsp/python-lsp-server/assets/6631116/e033fddd-2e1b-4bad-a988-17c13eef5275)
![hover_preview_3](https://github.com/python-lsp/python-lsp-server/assets/6631116/faac06b7-9fef-4724-b1e7-07509dc52abe)

HedgehogCode committed:
Improve hover results

- Use jedi Script.help instead of infer - Ensures that docstrings of the
  definition are used if possible instead of docstrings of the type
- Show all possible signatures or types of the element hovered instead
  of just the first one that matches the name

tkrabel-db commented on pylsp/plugins/hover.py:
```diff
@@ -8,43 +8,81 @@
+
+        # TODO(Review)
+        # We could also concatenate all docstrings we find in the definitions
+        # I am agains this because
```
```suggestion
        # I am against this because
```

tkrabel-db commented on pylsp/plugins/hover.py:
```diff
@@ -8,43 +8,81 @@
+        # We could also concatenate all docstrings we find in the definitions
+        # I am agains this because
+        # - If just one definition has a docstring, it gives a false impression of the hover element
+        # - If multiple definitions have a docstring, the user will probably not relize
```
```suggestion
        # - If multiple definitions have a docstring, the user will probably not realize
```

HedgehogCode committed:
Apply suggestions from code review (fix typos)

Co-authored-by: tkrabel-db <91616041+tkrabel-db@users.noreply.github.com>

PeterCardenas commented:
@tkrabel-db @HedgehogCode any way we can get this merged? looking forward to this change

tkrabel-db commented:
@PeterCardenas I'm afraid I have no rights to do so.
@ccordoba12 can you TAL?

tbung commented:
Thanks for your work on this! I am really looking forward for this to get merged.

I think there is still some room for improvement here, though. For variables with `None` in their union type or `Optional` type, the variable gets simply shown as `NoneType`. jedi-language-server is slightly better with showing the complete definition line in that regard.

Here is what python-lsp-server with this PR shows (without this PR python-lsp-server shows nothing at all in this situation):
![Screenshot from 2024-01-14 03-09-21](https://github.com/python-lsp/python-lsp-server/assets/4084982/0f9ec84d-9ca0-4d26-8340-e6cc90e3c160)

And here is what jedi-language-server shows:
![Screenshot from 2024-01-14 03-10-40](https://github.com/python-lsp/python-lsp-server/assets/4084982/26b25a75-b6ac-4686-87f2-5a84f30d7543)

The real solution would be to correctly show a union type here. Since both lsps don't do that I assume this is an issue with jedi, I haven't check though.

HedgehogCode committed:
Revise signatures and types on hover

- Only show signatures for function and class types
- Also show types if there are definitons without signatures next to
  definitions with signatures

HedgehogCode commented:
Sorry for the late reply.

The example above showed `NoneType()` because jedi returns a signature for this definition. However, signatures are only interesting for "function" and "class" ([BaseName#get_signatures](https://jedi.readthedocs.io/en/latest/docs/api-classes.html?highlight=BaseSignature#jedi.api.classes.BaseName.get_signatures)). I changed the code to only show signatures for "function" and "class".
The new code shows `Union[NoneType, int]` on hover (this was already implemented - we just did not get there because of the signature).

I also changed the code to show the type in cases where we have a signature but this is not the only option. Example:
```
def b():
    return 1

if input() == "yes":
    b = 10

b
```
On hover over "b":
```
b()
Union[b, int]
```

flexagoon commented:
Hi, what's the status of this? This issue is super annoying

ibrokemypie commented:
Anything that can be done to get this merged?
